File: apps/qizhi/mobile/MentorAI-Android/app/src/main/kotlin/com/mentorai/app/videoanalysis/VideoAPI.kt
package com.mentorai.app.videoanalysis

import android.content.Context
import android.net.Uri
import com.mentorai.app.chat.SSEMessage
import com.mentorai.app.chat.SSEStream
import com.mentorai.app.networking.APIClient
import com.mentorai.app.networking.AuthError
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File
import java.io.FileOutputStream
import java.io.InputStream
import java.util.UUID

/**
 * Mirrors the iOS `VideoAPI`. REST methods use `APIClient` helpers; chunked upload and the
 * Zhiyun import stream talk to OkHttp directly so we can pump multipart bodies and SSE.
 */
class VideoAPI(private val client: APIClient) {

    // ---- List + detail + operations ----

    suspend fun list(token: String): List<VideoSummary> =
        client.getEnvelope("/video/list", bearerToken = token)

    suspend fun detail(id: String, token: String): VideoDetail =
        client.getEnvelope("/video", mapOf("id" to id), token)

    /**
     * Kick off analysis for a video. `mode` selects 云端 (智云/超星) vs 本地 self-hosted model and
     * is sent as the `mode` query param (cloud|local), mirroring iOS `analyze(id:mode:token:)`.
     * Defaults to [AnalysisMode.Cloud]; param order (id, mode, token) matches iOS and the
     * detail-screen caller `analyze(d.id, mode, token)`.
     */
    suspend fun analyze(id: String, mode: AnalysisMode = AnalysisMode.Cloud, token: String) {
        // Returns ApiResponse[None]; we only care that it didn't error.
        val request = client.buildRequest(
            method = "GET",
            path = "/video/analyze",
            query = mapOf("id" to id, "mode" to mode.wireValue),
            bearerToken = token,
        )
        client.executeString(request)
    }

    /** Returns the new video id on Create, otherwise null. Mirrors iOS `VideoAPI.operate`. */
    suspend fun operate(params: VideoOperationRequest, token: String): String {
        val payload = client.json.encodeToString(VideoOperationRequest.serializer(), params)
        val request = client.buildRequest(
            method = "POST",
            path = "/video/operation",
            body = client.jsonBody(payload),
            bearerToken = token,
        )
        // The envelope's data is a string (the new id) on Create; otherwise null.
        val body = client.executeString(request)
        val env = client.json.parseToJsonElement(body) as? JsonObject
            ?: throw AuthError.Server(0, "无效的响应")
        if (env["success"]?.let { (it as? JsonPrimitive)?.booleanLiteral() } == false) {
            val msg = (env["error"] as? JsonPrimitive)?.contentOrNull
                ?: (env["message"] as? JsonPrimitive)?.contentOrNull
            throw AuthError.Server((env["code"] as? JsonPrimitive)?.intOrNull ?: 0, msg)
        }
        return (env["data"] as? JsonPrimitive)?.contentOrNull.orEmpty()
    }

    private fun JsonPrimitive.booleanLiteral(): Boolean? =
        if (isString) content.toBooleanStrictOrNull() else content.toBooleanStrictOrNull()

    // ---- Zhiyun (smart classroom) ----

    /**
     * The server returns groups; the iOS app flattens them to recording rows for display, and
     * applies the (optional) course-name filter client-side because the server dropped it.
     */
    suspend fun listZhiyunCourses(
        beginDate: String,
        endDate: String,
        courseName: String?,
        token: String,
    ): List<ZhiyunCourse> {
        val groups: List<ZhiyunCourseGroup> = client.getEnvelope(
            path = "/video/zhiyun/list",
            query = mapOf(
                "search_begin_date" to beginDate,
                "search_end_date" to endDate,
            ),
            bearerToken = token,
        )
        var rows = groups.flatMap { group ->
            group.items.map { item ->
                ZhiyunCourse(
                    courseId = group.courseId,
                    subId = item.subId,
                    courseName = group.courseName,
                    subTitle = item.subTitle,
                    teacherName = item.teacherName,
                    classBegin = item.classBegin,
                )
            }
        }
        val filter = courseName?.trim().orEmpty()
        if (filter.isNotEmpty()) {
            rows = rows.filter { it.courseName.contains(filter, ignoreCase = true) }
        }
        return rows
    }

    sealed class ZhiyunImportEvent {
        /** The server announces the new video id up-front via the `start` event: {"id": "..."}. */
        data class Start(val videoId: String) : ZhiyunImportEvent()
        data class Progress(val percent: Int) : ZhiyunImportEvent()
        data class Error(val message: String) : ZhiyunImportEvent()
        data class End(val videoId: String?) : ZhiyunImportEvent()
    }

    /**
     * GET /video/zhiyun/import?course_id=&sub_id=&import_id= — SSE with `start` + `loading` + `end`.
     * `importId` is echoed to the server so an out-of-band [cancelZhiyunImport] can stop the
     * download mid-flight (a cancelled client SSE alone is unreliable). Mirrors iOS
     * `importZhiyun(courseID:subID:importID:)`.
     */
    fun importZhiyun(courseId: String, subId: String, importId: String, token: String): Flow<ZhiyunImportEvent> = flow {
        val request = client.buildRequest(
            method = "GET",
            path = "/video/zhiyun/import",
            query = mapOf("course_id" to courseId, "sub_id" to subId, "import_id" to importId),
            bearerToken = token,
            accept = "text/event-stream",
        )
        SSEStream.messages(request, client.httpClient).collect { sse ->
            val event = parseZhiyunImport(sse) ?: return@collect
            emit(event)
        }
    }

    /**
     * 取消正在进行的智云课堂视频导入. Cancelling the client SSE alone is unreliable (proxies buffer
     * the upstream, so the backend only notices the disconnect after the download finished and the
     * video already landed in the DB — "取消后视频仍出现在列表里"). This writes an out-of-band cancel
     * marker so the backend stops at the next chunk and cleans up. Best-effort; mirrors iOS
     * `cancelZhiyunImport(importID:)`.
     */
    suspend fun cancelZhiyunImport(importId: String, token: String) {
        if (importId.isEmpty()) return
        val payload = buildJsonObject { put("import_id", importId) }.toString()
        val request = client.buildRequest(
            method = "POST",
            path = "/video/zhiyun/import/cancel",
            body = client.jsonBody(payload),
            bearerToken = token,
        )
        client.executeString(request)
    }

    private fun parseZhiyunImport(sse: SSEMessage): ZhiyunImportEvent? {
        val name = sse.event?.trim()?.lowercase()
        val payload = runCatching { client.json.parseToJsonElement(sse.data) }.getOrNull()
        val obj = payload as? JsonObject
        return when (name) {
            "start" -> {
                val id = (obj?.get("id") as? JsonPrimitive)?.contentOrNull?.takeIf { it.isNotEmpty() }
                    ?: return null
                ZhiyunImportEvent.Start(videoId = id)
            }
            "loading" -> {
                val p = (obj?.get("progress") as? JsonPrimitive)?.let {
                    it.intOrNull ?: it.doubleOrNull?.toInt()
                } ?: return null
                ZhiyunImportEvent.Progress(p.coerceIn(0, 100))
            }
            "error" -> {
                val msg = (obj?.get("error") as? JsonPrimitive)?.contentOrNull
                    ?: (obj?.get("message") as? JsonPrimitive)?.contentOrNull
                    ?: sse.data.ifEmpty { "导入失败" }
                ZhiyunImportEvent.Error(msg)
            }
            "end", "done" -> {
                val id = (obj?.get("id") as? JsonPrimitive)?.contentOrNull?.takeIf { it.isNotEmpty() }
                ZhiyunImportEvent.End(videoId = id)
            }
            else -> null
        }
    }

    // ---- Chunked upload ----

    /** Initialize a chunked upload with N chunks; returns the `upload_id`. */
    suspend fun initUpload(totalChunks: Int, token: String): String {
        val form = "chunks=$totalChunks"
        val id: String = client.postFormEnvelope("/video/init", form, token)
        if (id.isBlank()) throw AuthError.Server(0, "初始化上传失败")
        return id
    }

    /** Upload one 5 MB chunk via multipart. Mirrors iOS `uploadChunk`. */
    suspend fun uploadChunk(
        uploadId: String,
        index: Int,
        data: ByteArray,
        filename: String,
        token: String,
    ) {
        val body = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart("upload_id", uploadId)
            .addFormDataPart("index", index.toString())
            .addFormDataPart(
                name = "file",
                filename = filename,
                body = data.toRequestBody("application/octet-stream".toMediaType()),
            )
            .build()
        val request = client.buildRequest(
            method = "POST",
            path = "/video/upload",
            body = body,
            bearerToken = token,
        )
        // Envelope with `data: null` on success — just check non-error.
        client.executeString(request)
    }

    /** Finish the upload, asking the server to merge chunks. Returns the video + cover path. */
    suspend fun finishUpload(uploadId: String, token: String): UploadFinishResult {
        val body = client.formBody("upload_id=$uploadId")
        val request = client.buildRequest(
            method = "POST",
            path = "/video/finish",
            body = body,
            bearerToken = token,
        )
        val responseBody = client.executeString(request)
        val root = client.json.parseToJsonElement(responseBody) as? JsonObject
            ?: throw AuthError.Server(0, "合并上传失败")
        val data = root["data"] as? JsonObject ?: throw AuthError.Server(0, "合并上传失败")
        val path = (data["path"] as? JsonPrimitive)?.contentOrNull.orEmpty()
        val cover = (data["cover_path"] as? JsonPrimitive)?.contentOrNull
        if (path.isBlank()) throw AuthError.Server(0, "合并上传失败")
        return UploadFinishResult(videoPath = path, coverPath = cover)
    }

    companion object {
        const val CHUNK_SIZE: Int = 5 * 1024 * 1024  // 5 MB, matches iOS + web

        /**
         * Copy a content URI into the app cache so we can stream it predictably from disk
         * (URIs may not be seekable). Caller is responsible for deleting the returned file.
         */
        suspend fun stageUpload(context: Context, uri: Uri, fallbackName: String): File =
            withContext(Dispatchers.IO) {
                val ext = fallbackName.substringAfterLast('.', "mp4")
                val out = File(context.cacheDir, "upload-${UUID.randomUUID()}.$ext")
                context.contentResolver.openInputStream(uri).use { input: InputStream? ->
                    requireNotNull(input) { "无法读取所选视频" }
                    FileOutputStream(out).use { sink -> input.copyTo(sink) }
                }
                if (!out.exists() || out.length() == 0L) {
                    throw AuthError.Transport("无法读取所选视频")
                }
                out
            }
    }
}

File: apps/qizhi/mobile/MentorAI-iOS/MentorAI/VideoAnalysis/VideoAPI.swift
import Foundation
import os

/// 视频分析方式。`cloud`=智云课堂(超星)云端分析（默认）；`local`=本地自建模型直接分析。
/// Mirrors the web `AnalysisMode` (`api/video.ts`); sent as the `mode` query param to /video/analyze.
enum AnalysisMode: String {
    case cloud
    case local
}

struct VideoAPI {
    let baseURL: URL
    let client: APIClient
    let streamSession: URLSession
    private let log = Logger(subsystem: "com.mentorai.app", category: "VideoAPI")
    static let chunkSize: Int = 5 * 1024 * 1024  // 5 MB — matches web client

    init(baseURL: URL) {
        self.baseURL = baseURL
        self.client = URLSessionAPIClient(baseURL: baseURL)
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 300
        config.timeoutIntervalForResource = 3600
        self.streamSession = URLSession(configuration: config)
    }

    // MARK: - List + Detail + Operation

    func list(token: String) async throws -> [VideoSummary] {
        try await client.get("/video/list", query: [], bearerToken: token)
    }

    func detail(id: String, token: String) async throws -> VideoDetail {
        try await client.get("/video",
                             query: [URLQueryItem(name: "id", value: id)],
                             bearerToken: token)
    }

    @discardableResult
    func operate(_ params: VideoOperationRequest, token: String) async throws -> String {
        let url = try makeURL(path: "video/operation")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONEncoder().encode(params)

        let (data, response) = try await streamSession.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw AuthError.transport("收到非 HTTP 响应。")
        }
        let bodyPreview = String(data: data.prefix(500), encoding: .utf8) ?? "<binary>"
        log.debug("POST video/operation op=\(params.operation.rawValue, privacy: .public) → \(http.statusCode, privacy: .public) \(bodyPreview, privacy: .public)")
        let decoder = JSONDecoder()
        guard (200..<300).contains(http.statusCode) else {
            if let env = try? decoder.decode(APIEnvelope<String>.self, from: data),
               let msg = env.errorMessage {
                throw AuthError.server(status: http.statusCode, message: msg)
            }
            throw AuthError.server(status: http.statusCode, message: bodyPreview)
        }
        let envelope = try decoder.decode(APIEnvelope<String>.self, from: data)
        if !envelope.isSuccess {
            throw AuthError.server(status: envelope.code ?? http.statusCode,
                                   message: envelope.errorMessage ?? "操作失败")
        }
        return envelope.data ?? ""
    }

    func analyze(id: String, mode: AnalysisMode = .cloud, token: String) async throws {
        let _: EmptyDecodable = try await client.get(
            "/video/analyze",
            query: [
                URLQueryItem(name: "id", value: id),
                URLQueryItem(name: "mode", value: mode.rawValue),
            ],
            bearerToken: token
        )
    }

    // MARK: - Chunked Upload (init → upload → finish)

    struct UploadFinishResult {
        let videoPath: String
        let coverPath: String?
    }

    func initUpload(totalChunks: Int, token: String) async throws -> String {
        let url = try makeURL(path: "video/init")
        let body = "chunks=\(totalChunks)".data(using: .utf8) ?? Data()
        let envelope: APIEnvelope<String> = try await postForm(url: url, body: body, token: token)
        guard envelope.isSuccess, let uploadID = envelope.data, !uploadID.isEmpty else {
            throw AuthError.server(status: envelope.code ?? 500,
                                   message: envelope.errorMessage ?? "初始化上传失败")
        }
        return uploadID
    }

    func uploadChunk(uploadID: String, index: Int, data: Data, filename: String, token: String) async throws {
        let url = try makeURL(path: "video/upload")
        let boundary = "Boundary-\(UUID().uuidString)"
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.timeoutInterval = 300
        request.httpBody = multipartBody(
            boundary: boundary,
            fields: ["upload_id": uploadID, "index": "\(index)"],
            file: (name: "file", filename: filename, mimeType: "application/octet-stream", data: data)
        )

        let (respData, response) = try await streamSession.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw AuthError.transport("收到非 HTTP 响应。")
        }
        guard (200..<300).contains(http.statusCode) else {
            let preview = String(data: respData.prefix(500), encoding: .utf8) ?? ""
            throw AuthError.server(status: http.statusCode, message: preview)
        }
        let envelope = try JSONDecoder().decode(APIEnvelope<EmptyDecodable>.self, from: respData)
        if !envelope.isSuccess {
            throw AuthError.server(status: envelope.code ?? http.statusCode,
                                   message: envelope.errorMessage ?? "分片上传失败")
        }
    }

    func finishUpload(uploadID: String, token: String) async throws -> UploadFinishResult {
        let url = try makeURL(path: "video/finish")
        let body = "upload_id=\(uploadID)".data(using: .utf8) ?? Data()
        let envelope: APIEnvelope<[String: String]> = try await postForm(url: url, body: body, token: token)
        guard envelope.isSuccess, let dict = envelope.data, let path = dict["path"], !path.isEmpty else {
            throw AuthError.server(status: envelope.code ?? 500,
                                   message: envelope.errorMessage ?? "合并上传失败")
        }
        return UploadFinishResult(videoPath: path, coverPath: dict["cover_path"])
    }

    // MARK: - Zhiyun (smart classroom)

    /// `/video/zhiyun/list` now returns courses grouped (`[ZhiyunCourseGroup]`) and only
    /// supports date-range filtering server-side, so we flatten the groups into recording
    /// rows and apply the optional course-name filter on the client.
    func listZhiyunCourses(beginDate: String,
                          endDate: String,
                          courseName: String?,
                          token: String) async throws -> [ZhiyunCourse] {
        let query: [URLQueryItem] = [
            URLQueryItem(name: "search_begin_date", value: beginDate),
            URLQueryItem(name: "search_end_date", value: endDate),
        ]
        let groups: [ZhiyunCourseGroup] = try await client.get(
            "/video/zhiyun/list", query: query, bearerToken: token
        )
        var rows = groups.flatMap { group in
            group.items.map { item in
                ZhiyunCourse(
                    courseID: group.courseID,
                    subID: item.subID,
                    courseName: group.courseName,
                    subTitle: item.subTitle,
                    teacherName: item.teacherName,
                    classBegin: item.classBegin
                )
            }
        }
        if let courseName = courseName?.trimmingCharacters(in: .whitespacesAndNewlines), !courseName.isEmpty {
            rows = rows.filter { $0.courseName.localizedCaseInsensitiveContains(courseName) }
        }
        return rows
    }

    enum ZhiyunImportEvent: Equatable {
        case start(videoID: String)
        case progress(percent: Int)
        case error(String)
        case end(videoID: String?)
    }

    func importZhiyun(courseID: String, subID: String, importID: String, token: String) -> AsyncThrowingStream<ZhiyunImportEvent, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    let url = try makeURL(path: "video/zhiyun/import",
                                          query: [
                                            URLQueryItem(name: "course_id", value: courseID),
                                            URLQueryItem(name: "sub_id", value: subID),
                                            // 本次导入标识：配合 cancelZhiyunImport 让后端可中途取消
                                            URLQueryItem(name: "import_id", value: importID),
                                          ])
                    var request = URLRequest(url: url)
                    request.httpMethod = "GET"
                    request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
                    request.setValue("no-cache", forHTTPHeaderField: "Cache-Control")
                    request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
                    request.timeoutInterval = 300

                    log.debug("GET \(url.absoluteString, privacy: .public)")

                    for try await sse in SSEStream.messages(for: request, session: streamSession) {
                        if Task.isCancelled { break }
                        guard let event = Self.parseZhiyunEvent(sse) else { continue }
                        continuation.yield(event)
                        if case .end = event { break }
                        if case .error = event { break }
                    }
                    continuation.finish()
                } catch is CancellationError {
                    continuation.finish(throwing: CancellationError())
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    /// 取消正在进行的智云课堂视频导入。
    ///
    /// 仅取消客户端 SSE 任务并不可靠：反向代理常缓冲上游连接，后端要到下载完成、
    /// 视频已落库后才察觉断开，于是「取消后视频仍出现在列表里」。本接口写一个带外
    /// 取消标记，后端在下一个分片处即停止下载并清理，不落库。best-effort。
    func cancelZhiyunImport(importID: String, token: String) async throws {
        guard !importID.isEmpty else { return }
        let url = try makeURL(path: "video/zhiyun/import/cancel")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONEncoder().encode(["import_id": importID])

        let (data, response) = try await streamSession.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw AuthError.transport("收到非 HTTP 响应。")
        }
        let bodyPreview = String(data: data.prefix(300), encoding: .utf8) ?? "<binary>"
        log.debug("POST video/zhiyun/import/cancel → \(http.statusCode, privacy: .public) \(bodyPreview, privacy: .public)")
        guard (200..<300).contains(http.statusCode) else {
            throw AuthError.server(status: http.statusCode, message: bodyPreview)
        }
    }

    private static func parseZhiyunEvent(_ sse: SSEMessage) -> ZhiyunImportEvent? {
        let event = (sse.event ?? "").lowercased()
        let json = (try? JSONSerialization.jsonObject(with: Data(sse.data.utf8), options: [.fragmentsAllowed])) as? [String: Any]
        switch event {
        case "start":
            if let id = json?["id"] as? String, !id.isEmpty {
                return .start(videoID: id)
            }
            return nil
        case "loading":
            if let p = json?["progress"] as? Int {
                return .progress(percent: p)
            }
            if let p = json?["progress"] as? Double {
                return .progress(percent: Int(p))
            }
            return nil
        case "error":
            let msg = (json?["error"] as? String)
                ?? (json?["message"] as? String)
                ?? sse.data
            return .error(msg.isEmpty ? "导入失败" : msg)
        case "end", "done":
            // The server delivers the new video id in the end event: {"id": "..."}.
            let id = json?["id"] as? String
            return .end(videoID: (id?.isEmpty == false) ? id : nil)
        default:
            return nil
        }
    }

    // MARK: - Helpers

    private func postForm<T: Decodable>(url: URL, body: Data, token: String) async throws -> APIEnvelope<T> {
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.httpBody = body

        let (data, response) = try await streamSession.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw AuthError.transport("收到非 HTTP 响应。")
        }
        let preview = String(data: data.prefix(500), encoding: .utf8) ?? ""
        log.debug("POST \(url.absoluteString, privacy: .public) → \(http.statusCode, privacy: .public) \(preview, privacy: .public)")
        guard (200..<300).contains(http.statusCode) else {
            throw AuthError.server(status: http.statusCode, message: preview)
        }
        return try JSONDecoder().decode(APIEnvelope<T>.self, from: data)
    }

    private func multipartBody(boundary: String,
                               fields: [String: String],
                               file: (name: String, filename: String, mimeType: String, data: Data)) -> Data {
        var body = Data()
        let crlf = "\r\n"
        for (k, v) in fields {
            body.append("--\(boundary)\(crlf)".data(using: .utf8)!)
            body.append("Content-Disposition: form-data; name=\"\(k)\"\(crlf)\(crlf)".data(using: .utf8)!)
            body.append("\(v)\(crlf)".data(using: .utf8)!)
        }
        body.append("--\(boundary)\(crlf)".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"\(file.name)\"; filename=\"\(file.filename)\"\(crlf)".data(using: .utf8)!)
        body.append("Content-Type: \(file.mimeType)\(crlf)\(crlf)".data(using: .utf8)!)
        body.append(file.data)
        body.append(crlf.data(using: .utf8)!)
        body.append("--\(boundary)--\(crlf)".data(using: .utf8)!)
        return body
    }

    private func makeURL(path: String, query: [URLQueryItem] = []) throws -> URL {
        let trimmed = path.hasPrefix("/") ? String(path.dropFirst()) : path
        let base = baseURL.absoluteString.hasSuffix("/") ? baseURL.absoluteString : baseURL.absoluteString + "/"
        guard var components = URLComponents(string: base + trimmed) else {
            throw AuthError.transport("URL 组件无效：\(path)。")
        }
        if !query.isEmpty {
            components.queryItems = (components.queryItems ?? []) + query
        }
        guard let url = components.url else {
            throw AuthError.transport("无法为路径 \(path) 构建 URL。")
        }
        return url
    }
}
